plugins/csr/fu-csr-device.c

```diff
@@ -0,0 +1,613 @@
+	FuCsrDeviceQuirks	 quirks;
+	DfuState		 dfu_state;
+	guint32			 dnload_timeout;
+};
+
+G_DEFINE_TYPE (FuCsrDevice, fu_csr_device, FU_TYPE_USB_DEVICE)
+
+#define HID_REPORT_GET				0x01
+#define HID_REPORT_SET				0x09
+
+#define HID_REPORT_TYPE_INPUT			0x01
+#define HID_REPORT_TYPE_OUTPUT			0x02
+#define HID_REPORT_TYPE_FEATURE			0x03
+
+#define HID_FEATURE				0x0300
+
+#define FU_CSR_REPORT_ID_COMMAND		0x01
+#define FU_CSR_REPORT_ID_STATUS			0x02
+#define FU_CSR_REPORT_ID_CONTROL		0x03
+
+#define FU_CSR_COMMAND_HEADER_SIZE		6	/* bytes */
+#define FU_CSR_COMMAND_UPGRADE			0x01
+
+#define FU_CSR_STATUS_HEADER_SIZE		7
+
+#define FU_CSR_CONTROL_HEADER_SIZE		2	/* bytes */
+#define FU_CSR_CONTROL_CLEAR_STATUS		0x04
+#define FU_CSR_CONTROL_RESET			0xff
+
+/* maximum firmware packet, including the command header */
+#define FU_CSR_PACKET_DATA_SIZE			1023	/* bytes */
+
+#define FU_CSR_DEVICE_TIMEOUT			5000	/* ms */
+
+static void
+fu_csr_device_to_string (FuDevice *device, GString *str)
+{
+	FuCsrDevice *self = FU_CSR_DEVICE (device);
+	g_string_append (str, "  DfuCsrDevice:\n");
+	g_string_append_printf (str, "    state:\t\t%s\n", dfu_state_to_string (self->dfu_state));
+	g_string_append_printf (str, "    timeout:\t\t%" G_GUINT32_FORMAT "\n", self->dnload_timeout);
+}
+
+static void
+fu_csr_device_dump (const gchar *title, const guint8 *buf, gsize sz)
+{
+	if (g_getenv ("FWUPD_CSR_VERBOSE") == NULL)
+		return;
+	g_print ("%s (%" G_GSIZE_FORMAT "):\n", title, sz);
+	for (gsize i = 0; i < sz; i++)
+		g_print ("%02x ", buf[i]);
+	g_print ("\n");
+}
+
+void
+fu_csr_device_set_quirks (FuCsrDevice *self, FuCsrDeviceQuirks quirks)
+{
+	self->quirks = quirks;
+}
+
+gboolean
+fu_csr_device_attach (FuCsrDevice *self, GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (FU_USB_DEVICE (self));
+	gsize sz = 0;
+	guint8 buf[] = { FU_CSR_REPORT_ID_CONTROL, FU_CSR_CONTROL_RESET };
+
+	fu_csr_device_dump ("Reset", buf, sz);
+	if (!g_usb_device_control_transfer (usb_device,
+					    G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					    G_USB_DEVICE_REQUEST_TYPE_CLASS,
+					    G_USB_DEVICE_RECIPIENT_INTERFACE,
+					    HID_REPORT_SET,				/* bRequest */
+					    HID_FEATURE | FU_CSR_REPORT_ID_CONTROL,	/* wValue */
+					    0x0000,					/* wIndex */
+					    buf, sizeof(buf), &sz,
+					    FU_CSR_DEVICE_TIMEOUT, /* timeout */
+					    NULL, error)) {
+		g_prefix_error (error, "Failed to ClearStatus: ");
+		return FALSE;
+	}
+
+	/* check packet */
+	if (sz != FU_CSR_CONTROL_HEADER_SIZE) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "Reset packet was %" G_GSIZE_FORMAT " expected %i",
+			     sz, FU_CSR_CONTROL_HEADER_SIZE);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_csr_device_get_status (FuCsrDevice *self, GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (FU_USB_DEVICE (self));
+	gsize sz = 0;
+	guint8 buf[64];
+
+	/* hit hardware */
+	memset (buf, 0x00, sizeof(buf));
```

Comment:
fwupd is C99 right? If so `buf[64]  = {0}` should do the job ([c reference](http://en.cppreference.com/w/c/language/array_initialization))

Reply:
Agreed, thanks.